nemo/collections/asr/metrics/wer_bpe.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+import editdistance
+import torch
+from pytorch_lightning.metrics import TensorMetric
+
+from nemo.collections.common.tokenizers.tokenizer_spec import TokenizerSpec
+
+
+class WERBPE(TensorMetric):
+    """
+    This metric computes numerator and denominator for Overall Word Error Rate for BPE tokens (WER-BPE) between prediction and reference texts.
+    When doing distributed training/evaluation the result of res=WERBPE(predictions, targets, target_lengths) calls
+    will be all-reduced between all workers using SUM operations.
+    Here contains two numbers res=[wer_numerator, wer_denominator]. WERBPE=wer_numerator/wer_denominator.
+
+    If used with PytorchLightning LightningModule, include wer_numerator and wer_denominators inside validation_step results.
+    Then aggregate (sum) then at the end of validation epoch to correctly compute validation WER.
+
+    Example:
+       def validation_step(self, batch, batch_idx):
+           ...
+           wer_num, wer_denom = self.__wer(predictions, transcript, transcript_len)
+           return {'val_loss': loss_value, 'val_wer_num': wer_num, 'val_wer_denom': wer_denom}
+
+       def validation_epoch_end(self, outputs):
+           ...
+           wer_num = torch.stack([x['val_wer_num'] for x in outputs]).sum()
+           wer_denom = torch.stack([x['val_wer_denom'] for x in outputs]).sum()
+           tensorboard_logs = {'validation_loss': val_loss_mean, 'validation_avg_wer': wer_num / wer_denom}
+           return {'val_loss': val_loss_mean, 'log': tensorboard_logs}
+
+    Args:
+       vocabulary:
+       batch_dim_index:
+       use_cer:
+       ctc_decode:
+
+    Returns:
+       res: a torch.Tensor object with two elements: [wer_numerator, wer_denominators]. To correctly compute average
+       text word error rate, compute wer=wer_numerator/wer_denominators
+    """
+
+    def __init__(self, tokenizer: TokenizerSpec, batch_dim_index=0, use_cer=False, ctc_decode=True):
+        super().__init__(name="WER_BPE")
+        self.tokenizer = tokenizer
+        self.batch_dim_index = batch_dim_index
+        self.blank_id = tokenizer.tokenizer.vocab_size
```

Comment:
This is a strange line, is this correct?

Reply:
Not particularly. NeMo tokenizer (the first tokenizer) is wrapping an internal HF tokenizer, so this kind of dereferencing is unavoidable. And its an implementation detail hidden from user so it should be ok